### ISSUE TITLE
Thread blocked indefinetly in an MVar operation

### ISSUE THREAD

**Issue by PixeLInc** [aquarial/discord-haskell#1]
^ Title
Using GHC: 8.4.3

I get the error in the title when trying to run the example codes.

Code:
```haskell
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Control.Monad
import Control.Exception
import Data.Char (isSpace, toLower)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO

import Discord

main :: IO ()
main = do
    tok <- T.filter (not . isSpace) <$> TIO.readFile "auth-token.secret"
    dis <- loginRestGateway (Auth tok)

    finally (forever $ do
                e <- nextEvent dis
                case e of
                    MessageCreate m -> when (isPing (messageText m)) $ do
                        resp <- restCall dis (CreateMessage (messageChannel m) "Pong!" Nothing)
                        putStrLn $ show resp

                    _ -> pure ())
            (stopDiscord dis)

isPing :: T.Text -> Bool
isPing = T.isPrefixOf "ping" . T.map toLower
```

I'm slowly learning Haskell so I don't really know what's going on other than it's probably in the lib or something.
Lmk if there's any more info you need or files (like the stack or cabal)

**Comment by mattgreen** [comment]
I'm getting this as well with a valid auth token.

**Comment by mattgreen** [comment]
I figured it out! I was putting in the client secret for my app instead of the bot token. The bot token is listed under the bot section of your Discord app settings. I should know better, given that I already have a live bot.

The library needs to funnel back authentication errors instead of writing them to a log file, but hopefully this gets you unstuck.

**Comment by PixeLInc** [comment]
That's the thing, I'm 100% sure I'm using a valid auth token as well. It works on other bots. I'll look more into it I guess.

**Comment by PixeLInc** [comment]
Well, shit. I feel kinda stupid now.
I was prefixing my token with 'Bot' like how you normally would and it wasn't working but removing that lets it work fine.

Derp. Thanks for the tip.
(But also it should tell me the token is wrong instead of breaking but.. heh)

**Comment by aquarial** [comment]
I'm definitely less active than over the summer.

I just fixed the "Bot" prefix incompatibility by [only adding the prefix if it's not present](https://github.com/aquarial/discord-haskell/blob/974c3afad4c7e30db18d9421907fbebb9b2d6f59/src/Discord/Types/Prelude.hs#L26)

I agree it is much nicer when normal errors are propagated out into normal types. I'll work on changing the IO output of  `restCall`, `nextEvent`, and `readCache` to be `IO (Either ExceptionADT Output)`. Right now it's not clear from the types that `nextEvent` and `readCache` will fail in regular situations (like an invalid token), and `restCall`  uses `String` for error reporting.

**Comment by PixeLInc** [comment]
Alright. I've also been experiencing this error just randomly. It will be up and then go down at random times for this error as seen here:
![Error pic](https://i.imgur.com/3qv1Skx.png)

**Comment by aquarial** [comment]
That probably means the gateway thread is closing for some reason, maybe a parsing exception. I'll be able to work on this issue in a few hours.

I'm reopening because it's an ongoing issue.

**Comment by PixeLInc** [comment]
Alright. Thank you.

**Comment by aquarial** [comment]
I made a bunch of changes on the [dev branch](https://github.com/aquarial/discord-haskell/tree/dev), try using that and let me know what output you get. I'll merge it into master and bump the version to 5.1.0 if I don't find anything else broken and the dev branch helps you.

I'll remove the logging sometime since errors are reported back to the call site.

**Comment by PixeLInc** [comment]
Alright, I'll give it a shot and report back soon.

**Comment by aquarial** [comment]
Great to hear! I've merged these change into master and released v0.6.0

**Comment by t1m0thyj** [comment]
I created a Discord bot using the latest version of this library on the `master` branch, and got the same "thread blocked indefinitely" error after running it for a day:
> Event error: GatewayExceptionConnection (CloseRequest 1001 "") "Normal event loop close request"
> discord-haskell-bot-exe: thread blocked indefinitely in an MVar operation

I have another Discord bot that I've written in Python which occasionally outputs the following message:
> INFO:discord.gateway:Websocket closed with 1001 (CloudFlare WebSocket proxy restarting), attempting a reconnect.

Because of this it seems to me like a CloseRequest is apparently not too uncommon for Discord bots to encounter, and I would expect this library to attempt a reconnect automatically when it happens.
